Sparse and dense algorithms to do the matrix exponential

mingruyang opened the pull request:
[1] Added a Krylov subspace method to calculate exp(-tH)*v and exp(-itH)*v
[2] Added the irreducible Pade approximation to do the exponential of a general small dense matrix
[3] Added wrapper to calculate the norm of a matrix and do the LU factorization

To be done:
[1] Chebyshev approximation (the Chebyshev method can fail when solving some specific matrix, so I removed that implementation when I pushed).
[2] Treat cplx and real by using template for expPade.
[3] Check if the function handle the cplx type efficiently, e. g. when it is actually real, use algorithm for the real type.
[4] Add QR decomposition to treat the case where truncation not needed.

mingruyang committed:
[1] Added one-site and two-site TDVP algorithm

[2] Added a Krylov subspace method to calculate exp(-tH)*v and exp(-itH)*v
[3] Added the irreducible Pade approximation to do the exponential of a general small dense matrix
[4] Added wrapper to calculate the norm of matrix and do the LU factorization
[5] Added support of 0 and 1 center sites for localmpo and localop
[6] Corrected a bug in shift() of localmpo.h
[7] Added a sample code to use the tdvp function

To be done:
[1] Chebyshev approximation
[2] Use template for expPade
[3] Add an observer for TDVP
[4] Add QR decomposition to replace the SVD in one-site algorithm

mingruyang committed:
Delete unnecessary comment

mingruyang committed:
Correct the number of sites can be used in positionA

mingruyang committed:
Remove changes in localmpo.h localmposet.h localop.h all_mps.h .
Remove tdvp.h tdvp.cc and changes the Makefile accordingly.

mingruyang committed:
Remove compiled file.

mtfishman commented on itensor/iterativesolvers.h:
```diff
@@ -61,6 +61,22 @@ gmres(BigMatrixT const& A,
+// After finished, phi -> phi'.
+template <class BigMatrixT>
+Real
+krylov(BigMatrixT const& A,
```
Could we rename this as something more descriptive, like `applyExp`?

mtfishman commented on itensor/iterativesolvers.h:
```diff
@@ -937,6 +953,245 @@ arnoldi(const BigMatrixT& A,
+    auto maxiter_ = args.getSizeT("MaxIter",30);
+    auto errgoal_ = args.getReal("ErrGoal",1E-10);
+    auto debug_level_ = args.getInt("DebugLevel",-1);
+	int ideg_ = args.getInt("Ideg",6);
```
Could we make this argument name more descriptive, like `PadeApproxDeg`?

mingruyang replied:
Good suggestion. I have changed that.

mingruyang committed:
Changed the function name and argument name according to Matt's suggestion.

mtfishman commented:
What do you think of changing the convention to be to apply `exp(tH)*v` or `exp(itH)*v` instead of `exp(-tH)*v` or `exp(itH)*v` for an input `applyExp(H,v,t,...)`? 

That would be consistent with the `expHermitian` function in ITensor (I also prefer that when I write `applyExp(H,v,-0.1)` I know that I am doing `exp(-0.1*H)*v`, I think it reads a bit better).

mingruyang commented:
I checked this page: http://itensor.org/docs.cgi?vers=cppv3&page=formulas/tevol_mps_mpo <http://itensor.org/docs.cgi?vers=cppv3&page=formulas/tevol_mps_mpo>, and found the function toExpH(ampo,tau) use the convention exp(- tau H).
That’s why I use this convention.

Maybe we need change both applyExp and toExpH to be consistent with expHermitian?

> On Sep 25, 2019, at 12:53 PM, mtfishman <notifications@github.com> wrote:
> 
> What do you think of changing the convention to be to apply exp(tH)*v or exp(itH)*v instead of exp(-tH)*v or exp(itH)*v for an input applyExp(H,v,t,...)?
> 
> That would be consistent with the expHermitian function in ITensor (I also prefer that when I write applyExp(H,v,-0.1) I know that I am doing exp(-0.1*H)*v, I think it reads a bit better).
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/ITensor/ITensor/pull/267?email_source=notifications&email_token=AEHX2T7HQ63OS3JXJEOA2C3QLO6VNA5CNFSM4I2P4QCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7TENXQ#issuecomment-535185118>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AEHX2T4ATDAOEPTKTZVNXG3QLO6VNANCNFSM4I2P4QCA>.
>

mtfishman commented:
I think it is ok for now for `toExpH` to be inconsistent with `applyExp` and `expHermitian`. `toExpH` is a bit more physics oriented (since it works with `AutoMPO` and is mostly focused on time evolution, where `exp(-tH)` is often wanted), but `applyExp` and `expHermitian` are more general linear algebra functions, where we don't want to impose any specific conventions.

mtfishman commented:
Along those lines, I think it also makes sense to have people explicitly input an imaginary scale factor (i.e. input `applyExp(H,v,0.1_i)` to do `exp(i*0.1*H)*v`) instead of doing this through a keyword argument and imposing that the input scale is `Real`. The reason for this is that in theory, one may want to evolve by a general complex scale, but the current system does not allow that (but I believe the algorithm should be able to handle that more general case?).

mingruyang commented:
There are several layers.
I think for applyExp, the BigMatrixT type should be localmpo or localmposet, so the case when it will be used is more like the toExpH.
But for expMatrixApply, I think I need to make its convention to be the same as the expHermitian, since it is a general linear algebra for the Matrix type.
What do you think?

> On Sep 25, 2019, at 1:16 PM, mtfishman <notifications@github.com> wrote:
> 
> I think it is ok for now for toExpH to be inconsistent with applyExp and expHermitian. toExpH is a bit more physics oriented (since it works with AutoMPO and is mostly focused on time evolution, where exp(-tH) is often wanted), but applyExp and expHermitian are more general linear algebra functions, where we don't want to impose any specific conventions.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/ITensor/ITensor/pull/267?email_source=notifications&email_token=AEHX2TZUBPWV4PVQ2SGTLZ3QLPBJXA5CNFSM4I2P4QCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7TGVSI#issuecomment-535194313>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AEHX2T27FZEK4IYDL4N4IALQLPBJXANCNFSM4I2P4QCA>.
>

mingruyang commented:
Yes, the algorithm is able to handle the cplx t.

My primary reason for use a keyword argument instead is that when we do TDVP, the trotter error is usually much smaller than Zaletel’s method or TEBD as long as the bond dimension is enough big.
I thought the case we want to use the cplx t is to reduce the trotter error such as in Zaletel’s method, and I think it is not necessary for the TDVP.

But I agree with you that theoretically it should be able to deal with cplx t.
I will change that.

> On Sep 25, 2019, at 1:28 PM, mtfishman <notifications@github.com> wrote:
> 
> Along those lines, I think it also makes sense to have people explicitly input an imaginary scale factor (i.e. input applyExp(H,v,0.1_i) to do exp(i*0.1*H)*v) instead of doing this through a keyword argument and imposing that the input scale is Real. The reason for this is that in theory, one may want to evolve by a general complex scale, but the current system does not allow that (but I believe the algorithm should be able to handle that more general case?).
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub <https://github.com/ITensor/ITensor/pull/267?email_source=notifications&email_token=AEHX2TYL67ZBQELKP4TFWMDQLPCVFA5CNFSM4I2P4QCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7THZXY#issuecomment-535198943>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AEHX2T2DRN4AMQ7DHC5WC43QLPCVFANCNFSM4I2P4QCA>.
>

mtfishman commented:
`BigMatrixT` is not necessarily a `LocalOp` or anything related to physics. I would consider this functionality to be a fundamental linear algebra operation, so we should keep the interface as general as possible (i.e. imagine someone from math or machine learning using it, I think they would be very surprised by the convention that the scale factor is negated).

> There are several layers. I think for applyExp, the BigMatrixT type should be localmpo or localmposet, so the case when it will be used is more like the toExpH. But for expMatrixApply, I think I need to make its convention to be the same as the expHermitian, since it is a general linear algebra for the Matrix type. What do you think?
> […](#)
> On Sep 25, 2019, at 1:16 PM, mtfishman ***@***.***> wrote: I think it is ok for now for toExpH to be inconsistent with applyExp and expHermitian. toExpH is a bit more physics oriented (since it works with AutoMPO and is mostly focused on time evolution, where exp(-tH) is often wanted), but applyExp and expHermitian are more general linear algebra functions, where we don't want to impose any specific conventions. — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub <#267?email_source=notifications&email_token=AEHX2TZUBPWV4PVQ2SGTLZ3QLPBJXA5CNFSM4I2P4QCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7TGVSI#issuecomment-535194313>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AEHX2T27FZEK4IYDL4N4IALQLPBJXANCNFSM4I2P4QCA>.

mtfishman commented:
Thanks! It looks like that flag for whether or not it is complex is a bit baked in, you may be able to just change the input signature from `Real t` to `Cplx t` (though maybe that will make certain cases like applying the exponential of a real symmetric operator with a real scale factor to a real ITensor output a complex ITensor, perhaps that could be checked for as a special case).

> Yes, the algorithm is able to handle the cplx t. My primary reason for use a keyword argument instead is that when we do TDVP, the trotter error is usually much smaller than Zaletel’s method or TEBD as long as the bond dimension is enough big. I thought the case we want to use the cplx t is to reduce the trotter error such as in Zaletel’s method, and I think it is not necessary for the TDVP. But I agree with you that theoretically it should be able to deal with cplx t. I will change that.
> […](#)
> On Sep 25, 2019, at 1:28 PM, mtfishman ***@***.***> wrote: Along those lines, I think it also makes sense to have people explicitly input an imaginary scale factor (i.e. input applyExp(H,v,0.1_i) to do exp(i*0.1*H)*v) instead of doing this through a keyword argument and imposing that the input scale is Real. The reason for this is that in theory, one may want to evolve by a general complex scale, but the current system does not allow that (but I believe the algorithm should be able to handle that more general case?). — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub <#267?email_source=notifications&email_token=AEHX2TYL67ZBQELKP4TFWMDQLPCVFA5CNFSM4I2P4QCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7THZXY#issuecomment-535198943>, or mute the thread <https://github.com/notifications/unsubscribe-auth/AEHX2T2DRN4AMQ7DHC5WC43QLPCVFANCNFSM4I2P4QCA>.

mtfishman commented on itensor/tensor/algs_impl.h:
```diff
@@ -189,6 +190,122 @@ SVD(MatM && M,
+	else
+		{
+		//info = detail::chebyshevExp(y.data(),N,F.data());
+		Error("ideg cannot be less than 1");
```
This check could be done at the beginning of the function (just to make things a bit cleaner).

mtfishman replied:
Oh sorry, I didn't realize it was possibly a placeholder for using a different algorithm. But anyway, I think the algorithm choice should be done with a flag instead of chosen by whether or not `ideg` is a valid degree for the Padé approximation.

mingruyang replied:
I agree. But currently I didn't include the Chebyshev method, since I found it can be very unstable for some matrix when I tested early on. So I turned off the option of Chebyshev for now. And this is simply a check for the validity of "ideg" for the Pade approximation.

I don't know if you still want the Chebyshev choice. If yes, I can make a new flag for that.

mtfishman replied:
We can worry about that later, I am not familiar with these methods anyway :).

mtfishman commented:
So basically I think both `applyExp` and `expMatrixApply` should have the same sign convention as `ExpHermitian`.

> `BigMatrixT` is not necessarily a `LocalOp` or anything related to physics. I would consider this functionality to be a fundamental linear algebra operation, so we should keep the interface as general as possible (i.e. imagine someone from math or machine learning using it, I think they would be very surprised by the convention that the scale factor is negated).
> 
> > There are several layers. I think for applyExp, the BigMatrixT type should be localmpo or localmposet, so the case when it will be used is more like the toExpH. But for expMatrixApply, I think I need to make its convention to be the same as the expHermitian, since it is a general linear algebra for the Matrix type. What do you think?
> > […](#)
> > On Sep 25, 2019, at 1:16 PM, mtfishman _**@**_.***> wrote: I think it is ok for now for toExpH to be inconsistent with applyExp and expHermitian. toExpH is a bit more physics oriented (since it works with AutoMPO and is mostly focused on time evolution, where exp(-tH) is often wanted), but applyExp and expHermitian are more general linear algebra functions, where we don't want to impose any specific conventions. — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub <#267?email_source=notifications&email_token=AEHX2TZUBPWV4PVQ2SGTLZ3QLPBJXA5CNFSM4I2P4QCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7TGVSI#issuecomment-535194313>, or mute the thread https://github.com/notifications/unsubscribe-auth/AEHX2T27FZEK4IYDL4N4IALQLPBJXANCNFSM4I2P4QCA.

mingruyang commented:
I just realized I use a time-stepping strategy to make the Krylov algorithm more stable for different time step. In that strategy to adjust the time-step size, it appears to me that the time step should be real to have "t = abs(t) *sgn(t)". I'm not quite sure if the time step becomes cplx, using "t = abs(t) * exp(i arg(t))" still works for that strategy.

I also checked Jutho's Krylovkit.jl. In his implementation, the time step looks like also need to be real.

So maybe I'll keep the keyword argument for now and improve it in the future?

> Thanks! It looks like that flag for whether or not it is complex is a bit baked in, you may be able to just change the input signature from `Real t` to `Cplx t` (though maybe that will make certain cases like applying the exponential of a real symmetric operator with a real scale factor to a real ITensor output a complex ITensor, perhaps that could be checked for as a special case).
> 
> > Yes, the algorithm is able to handle the cplx t. My primary reason for use a keyword argument instead is that when we do TDVP, the trotter error is usually much smaller than Zaletel’s method or TEBD as long as the bond dimension is enough big. I thought the case we want to use the cplx t is to reduce the trotter error such as in Zaletel’s method, and I think it is not necessary for the TDVP. But I agree with you that theoretically it should be able to deal with cplx t. I will change that.
> > […](#)
> > On Sep 25, 2019, at 1:28 PM, mtfishman _**@**_.*__> wrote: Along those lines, I think it also makes sense to have people explicitly input an imaginary scale factor (i.e. input applyExp(H,v,0.1_i) to do exp(i_0.1_H)*v) instead of doing this through a keyword argument and imposing that the input scale is Real. The reason for this is that in theory, one may want to evolve by a general complex scale, but the current system does not allow that (but I believe the algorithm should be able to handle that more general case?). — You are receiving this because you authored the thread. Reply to this email directly, view it on GitHub <#267?email_source=notifications&email_token=AEHX2TYL67ZBQELKP4TFWMDQLPCVFA5CNFSM4I2P4QCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7THZXY#issuecomment-535198943>, or mute the thread https://github.com/notifications/unsubscribe-auth/AEHX2T2DRN4AMQ7DHC5WC43QLPCVFANCNFSM4I2P4QCA.

mtfishman commented:
Also I am a bit confused about something. In `expMatrixApply`, does it first compute the exponential of the input Matrix, and then apply it to the input vector (i.e. can these steps be decoupled)? 

It seems like maybe this could just be a function `exp(Matrix)` that exponentiates the matrix, and then multiplying it by the vector can be done after (i.e. in the `applyExp` code). I think this would be better, since then we could use the `exp` function with the Padé approximation in other places in ITensor (for example, we could use it to help with exponentiating dense ITensors that are not Hermitian).

If you don't want to deal with that in this pull request, that is fine (we can view it as an implementation detail of `applyExp` and think about making it more general later).

mingruyang commented:
Yes, in the expMatrixApply and padeExp, first we compute the exponential of the dense Matrix, and then apply it to the input vector.

I implemented it that way for historical reasons. Originally I want to also have the Chebyshev approximation and give a uniform interface for both the Chebyshev and Pade. In the Chebyshev way, it will be more efficient to get the result of exp(M)*v along the way instead of first getting the exp(M) and then applying it to v. But it's different for Pade.

So I think I will leave expMatrixApply as it has already been. But I will change the interface of padeExp to get only the exp(M).



> Also I am a bit confused about something. In `expMatrixApply`, does it first compute the exponential of the input Matrix, and then apply it to the input vector (i.e. can these steps be decoupled)?
> 
> It seems like maybe this could just be a function `exp(Matrix)` that exponentiates the matrix, and then multiplying it by the vector can be done after (i.e. in the `applyExp` code). I think this would be better, since then we could use the `exp` function with the Padé approximation in other places in ITensor (for example, we could use it to help with exponentiating dense ITensors that are not Hermitian).
> 
> If you don't want to deal with that in this pull request, that is fine (we can view it as an implementation detail of `applyExp` and think about making it more general later).

mtfishman commented:
Oh I see, I didn't appreciate the fact that you have some analysis at the beginning to calculate an effective timestep. Though I don't quite see why it would work with pure real and pure imaginary `t` but not for general complex? Should it not just depend on the norm of `t`? Anyway, if it is not obvious right now, I think it is still better to allow a general `Cplx t` input, and then just have a quick check at the beginning that the input is purely real or imaginary and deal with the more general case later.

It looks like KrylovKit allows for general complex scale factors (https://github.com/Jutho/KrylovKit.jl/blob/fd50d243c0f55810c07561ba75d3557470a92745/test/expintegrator.jl#L28).

> I just realized I use a time-stepping strategy to make the Krylov algorithm more stable for different time step. In that strategy to adjust the time-step size, it appears to me that the time step should be real to have "t = abs(t) *sgn(t)". I'm not quite sure if the time step becomes cplx, using "t = abs(t) * exp(i arg(t))" still works for that strategy.

mtfishman commented:
I see, that makes sense. I think changing the interface for `padeExp` is the right way to go here, that is one that we probably want to use elsewhere in ITensor (and we could just wrap that as `exp`).

> Yes, in the expMatrixApply and padeExp, first we compute the exponential of the dense Matrix, and then apply it to the input vector.
> 
> I implemented it that way for historical reasons. Originally I want to also have the Chebyshev approximation and give a uniform interface for both the Chebyshev and Pade. In the Chebyshev way, it will be more efficient to get the result of exp(M)*v along the way instead of first getting the exp(M) and then applying it to v. But it's different for Pade.
> 
> So I think I will leave expMatrixApply as it has already been. But I will change the interface of padeExp to get only the exp(M).

mingruyang commented:
I'm now confused about Jutho's KrylovKit.jl. It looks like in the interface of the function expintegrator, t is allowed to be cplx. But in line https://github.com/Jutho/KrylovKit.jl/blob/fd50d243c0f55810c07561ba75d3557470a92745/src/matrixfun/expintegrator.jl#171, It appears to me that sgn*Δτ doesn't retrieve the original t if t is cplx number.

What do you think?

Anyway, I will first test the applicability of cplx t for the time-stepping strategy and then change the interface to allow cplx t.

mtfishman commented:
It looks like he first splits off the norm and the phase of the time step:

https://github.com/Jutho/KrylovKit.jl/blob/fd50d243c0f55810c07561ba75d3557470a92745/src/matrixfun/expintegrator.jl#L106

and uses the original phase throughout the algorithm. It looks like the norm of the original time step is at least used here:

https://github.com/Jutho/KrylovKit.jl/blob/fd50d243c0f55810c07561ba75d3557470a92745/src/matrixfun/expintegrator.jl#L159

and then some other mechanisms are used for finding a new time step norm. So indeed it looks like the phase of the time step can be factored out as a constant (as we might expect), and the norm is modified.
 
> I'm now confused about Jutho's KrylovKit.jl. It looks like in the interface of the function expintegrator, t is allowed to be cplx. But in line https://github.com/Jutho/KrylovKit.jl/blob/fd50d243c0f55810c07561ba75d3557470a92745/src/matrixfun/expintegrator.jl#171, It appears to me that sgn*Δτ doesn't retrieve the original t if t is cplx number.
> 
> What do you think?
> 
> Anyway, I will first test the applicability of cplx t for the time-stepping strategy and then change the interface to allow cplx t.

mingruyang commented:
I see.  I didn't realize that the sign() function in Julia is to get the phase when the input is complex, not just sign in the sense of real numbers :P

mtfishman commented:
Ah yeah I could see how that would be confusing. I think they are using the general definition `sign(x) = x/|x|`.

mtfishman commented:
Hey Mingru,

I added a test for `applyExp`. In the process, I found a small bug in `expMatrixApply` that only was caught in debug mode (the output of the LAPACK norm function wasn't being captured, triggering an error).

I have another small question about the interface. One of the arguments is `"UseLanczos"`. Is the idea that the algorithm uses Lanczos if the input `BigMatrixT` is Hermitian, and otherwise it uses Arnoldi? Could we instead change the name of that argument to be `"IsHermitian"`? I think that is a bit clearer for people (since they might not know how Lanczos or Arnoldi are being used in this algorithm). 

I haven't done thorough testing, but over all this looks good! So I think once you make the following changes:

1. Change the interface to `applyExp` to accept a general `Cplx` timestep (and remove the "IsRealTevol" argument)
2. Change the interface to `padeExp` to just output the exponentiated matrix (this one is less important for the current pull request)
3. Change the name of "UseLanczos" to "IsHermitian" (?)

this should be good to go! Thanks for the contribution, this will be a nice addition to ITensor.

Cheers,
Matt

mingruyang commented:
Hi Matt,

I'm now working on your suggestions.

As for the argument "UseLanczos", it's not equivalent to that the matrix is
Hermitian.

Yes, if the matrix is Hermitian, we can use the Lanczos, but the
orthogonality of the basis might quickly start to be lost at some point
because of round-off errors. Also, because of the machine accuracy, the
matrix which is Hermitian theoretically might not be exact Hermitian in
implementation. So the Lanczos is not that stable. But as far as what I
have tested, when the matrix is theoretically Hermitian, I didn't get
larger errors when using Lanczos than using Arnoldi. If you think we don't
need to worry about this issue, then I can change the argument name.

On Thu, Sep 26, 2019 at 11:33 AM mtfishman <notifications@github.com> wrote:

> Hey Mingru,
>
> I added a test for applyExp. In the process, I found a small bug in
> expMatrixApply that only was caught in debug mode (the output of the
> LAPACK norm function wasn't being captured, triggering an error).
>
> I have another small question about the interface. One of the arguments is
> "UseLanczos". Is the idea that the algorithm uses Lanczos if the input
> BigMatrixT is Hermitian, and otherwise it uses Arnoldi? Could we instead
> change the name of that argument to be "IsHermitian"? I think that is a
> bit clearer for people (since they might not know how Lanczos or Arnoldi
> are being used in this algorithm).
>
> I haven't done thorough testing, but over all this looks good! So I think
> once you make the following changes:
>
>    1. Change the interface to applyExp to accept a general Cplx timestep
>    (and remove the "IsRealTevol" argument)
>    2. Change the interface to padeExp to just output the exponentiated
>    matrix (this one is less important for the current pull request)
>    3. Change the name of "UseLanczos" to "IsHermitian" (?)
>
> this should be good to go! Thanks for the contribution, this will be a
> nice addition to ITensor.
>
> Cheers,
> Matt
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ITensor/ITensor/pull/267?email_source=notifications&email_token=AEHX2T77KITZ7USPZAFSZ23QLT5WFA5CNFSM4I2P4QCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7WRFQY#issuecomment-535630531>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEHX2T5R74HPY4ZKZD3ZKUTQLT5WFANCNFSM4I2P4QCA>
> .
>


-- 
*Mingru Yang*
Research Assistant
*Department of Physics & Astronomy*
University of California, Irvine
CA 92697

mtfishman commented:
The "IsHermitian" flag would be off by default, so the function would still call Arnoldi by default. Then, if someone knows their operator is supposed to be Hermitian, they can try {"IsHermitian=",true} to see if that speeds up their calculation. It is only a name change since people may not be aware of the distinction between `Arnoldi` and `Lanczos`, especially in this context.

mtfishman commented:
It is similar to how in Julia, if you don't put an exactly symmetric matrix into `eigs`, you specify if your matrix is Hermitian by using the `Hermitian` wrapper:

```julia
using Arpack

A = randn(5,5) |> x->x+x'+eps()*randn(size(x))

ishermitian(A) # false

eigs(A)   # calls Arpack
eigs(Hermitain(A))  # call Lanczos
```

The point is just to not expose too much of the implementation detail to the user.

mingruyang commented:
I see. Sounds good!

On Fri, Sep 27, 2019, 11:41 AM mtfishman <notifications@github.com> wrote:

> It is similar to how in Julia, if you don't put an exactly symmetric
> matrix into eigs, you specify if your matrix is Hermitian by using the
> Hermitian wrapper:
>
> using Arpack
>
> A = randn(5,5) |> x->x+x'+eps()*randn(size(A))
> ishermitian(A) # false
> eigs(A)   # calls Arpackeigs(Hermitain(A))  # call Lanczos
>
> The point is just to not expose too much of the implementation detail to
> the user.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ITensor/ITensor/pull/267?email_source=notifications&email_token=AEHX2T2J4ZXDSSLUH7L7ILLQLZHTFA5CNFSM4I2P4QCKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7ZYQYQ#issuecomment-536053858>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEHX2T2XDGLKJJPPTEHBUWLQLZHTFANCNFSM4I2P4QCA>
> .
>

mingruyang committed:
[1] Corrected sign convention
[2] Allow cplx t
[3] Changed some argument name
[4] padeExp now only do matrix exponential
[5] Use &= to do matrix or vector copy

mingruyang commented:
Hi Matt,

I integrated your commits and modified according to your suggestions. Small modifications are made in the unittest to test the complex time step. it passed test.

Best,
Mingru

mtfishman commented:
Thanks for making the changes and thanks again for the contribution! I will merge this.

When you are ready, could you make a pull request for your work generalizing `LocalOp`/`LocalMPO` to more general unit cell sizes? Then we can discuss the best way to incorporate your TDVP code.
